okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * (c) Copyright 2017 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting3.okhttp;
+
+import com.google.common.util.concurrent.SettableFuture;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.remoting.api.errors.QosException;
+import com.palantir.remoting.api.errors.RemoteException;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.HttpUrl;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class RemotingOkHttpCall extends ForwardingCall {
+
+    private static final Logger log = LoggerFactory.getLogger(RemotingOkHttpCall.class);
+
+    private static final ResponseHandler<RemoteException> remoteExceptionHandler =
+            RemoteExceptionResponseHandler.INSTANCE;
+    private static final ResponseHandler<IOException> catchAllExceptionHandler = IoExceptionResponseHandler.INSTANCE;
+    private static final ResponseHandler<QosException> qosHandler = QosExceptionResponseHandler.INSTANCE;
+
+    private final BackoffStrategy backoffStrategy;
+    private final UrlSelector urls;
+    private final RemotingOkHttpClient client;
+    private final ScheduledExecutorService schedulingExecutor;
+    private final ExecutorService executionExecutor;
+    private final Duration timeout;
```

Comment:
might be worth elaborating on the purpose of this timeout as there are so many timeouts involved with http rpc already

Reply:
added.

Reply:
done